starport/pkg/ctxreader/ctxreader.go

```diff
@@ -18,18 +24,21 @@ func New(ctx context.Context, r io.Reader) io.Reader {
 }
 
 func (r *cancelableReader) Read(data []byte) (n int, err error) {
```

Comment:
There is also another case that this race cond can happen when you call `cancelableReader.Read` in different goroutines. To solve both I suggest something like below:

Btw, we don't need to create the goroutine or call `r.Reader.Read` if context is already canceled or previous read returned with an error.

```go
type cancelableReader struct {
	io.Reader
	ctx context.Context
	m sync.Mutex
	err error
}

func (r *cancelableReader) Read(data []byte) (n int, err error) {
	r.m.Lock()
	defer r.m.Unlock()

	if r.err != nil {
		return 0, r.err
	}

	var un int
	var uerr error

	isRead := make(chan struct{})

	go func() {
		un, uerr = r.Reader.Read(data)
		close(isRead)
	}()

	select {
	case <-r.ctx.Done():
		r.err = r.ctx.Err()
		return 0, r.ctx.Err()

	case <-isRead:
		r.err = uerr
		return un, uerr
	}
}
```